### ISSUE TITLE
Fully implement X509Store and related APIs

### ISSUE THREAD

**Issue by baulig** [mono/mono#7019]
### Overview

Our X509Store implementation needs to be cleaned up and finished.

Internally, we are using three different mechanisms for certificate validation:

1. AppleTls uses the native macOS / iOS certificate store.
1. BTLS uses its own store in `~/.config/.mono/new-certs`.
1. Legacy (which will eventually be removed) uses its own store in `~/.config/.mono/certs` (which is incompatible with the BTLS store).

Ideally, these APIs should be integrated with our TLS Provider selection mechanism and use a provider-specific backend.  This would allow us to use the native certificate store on a few platforms as a backend.

The X509Store APIs cannot access 1.) and accessing 2.) is only possible via internal APIs.

### Related issues:

- We also need to decide where to store system-wide configuration data on OS X since `/usr/share` is not writeable.
- People ask about adding an environment variable.
- On Ubuntu and some other Linux distros, evaluate whether we can use the certificates that are shipped with the distribution in for instance `/etc/ssl/certs`.
- Evaluate whether we can use the native certificate store on Windows.

### X509Store versus Client Certificates
There has been a slight confusion about what this issue is about and what it is not.

This issue (#7019) is about the X509Store API and the underlying certificate store.  It is important to understand that these APIs are **not** used by the TLS Provider to *validate* certificates.  And they're also not used to *provide* certificates for authentication.

The *certificate validation* code accesses the underlying certificate store directly via native code.

There is a Mono-specific API [MonoTlsSettings.TrustAnchors](https://github.com/mono/mono/blob/master/mcs/class/Mono.Security/Mono.Security.Interface/MonoTlsSettings.cs#L77) which can be used to add trusted certificates - either on a per-connection or on a per-app basis - but it does not go through the `X509Store`. 

Connecting the `MonoTlsSettings.TrustAnchors` mechanism with the `X509Store` is something we should evaluate.

When using `HttpListener` (the only way you can currently provide a certificate for authentication), you need to explicitly pass it the certificate that it should use.  It will not use `X509Store` or any similar mechanism.

So to summarize, this issue deals with how and where certificates should be stored and how to access and manipulate them with a client-side API.

Supporting Client Certificates (#7075) goes far beyond just *providing* certificates and somehow fetching them from a store.  The major challenge for client certificate support are the changes to the TLS Handshake and the different semantics for established connections.  Feeding a certificate to the provider can - at least for the initial implementation - be done similarly how we're currently doing it for `HttpListener`.

**Comment by baulig** [comment]
We also need to decide where to store system-wide configuration data on OS X since `/usr/share` is not writeable.

**Comment by baulig** [comment]
Changing title back, this has nothing to do with client certificates.

**Comment by baulig** [comment]
The epic label is also wrong and should be changed.
